File: khalti-android/src/main/java/com/khalti/checkOut/Wallet/WalletPresenter.java
package com.khalti.checkOut.Wallet;

import androidx.annotation.NonNull;

import java.util.HashMap;

import com.khalti.checkOut.Wallet.helper.WalletConfirmPojo;
import com.khalti.checkOut.Wallet.helper.WalletInitPojo;
import com.khalti.checkOut.helper.Config;
import com.khalti.checkOut.api.ErrorAction;
import com.khalti.checkOut.api.OnCheckOutListener;
import com.khalti.utils.Constant;
import com.khalti.utils.EmptyUtil;
import com.khalti.utils.GuavaUtil;
import com.khalti.utils.HtmlUtil;
import com.khalti.utils.NumberUtil;
import com.khalti.utils.Store;
import com.khalti.utils.StringUtil;
import com.khalti.utils.ValidationUtil;

import rx.Observable;
import rx.Subscriber;
import rx.subscriptions.CompositeSubscription;

public class WalletPresenter implements WalletContract.Presenter {
    @NonNull
    private final WalletContract.View view;
    private WalletModel model;
    private Config config;
    private CompositeSubscription compositeSubscription;

    private String pinWebLink;
    private int clicks = 0;

    public WalletPresenter(@NonNull WalletContract.View view) {
        this.view = GuavaUtil.checkNotNull(view);
        view.setPresenter(this);
        model = new WalletModel();
        compositeSubscription = new CompositeSubscription();
    }

    @Override
    public void onCreate() {
        config = Store.getConfig();
        if (EmptyUtil.isNotNull(config.getMobile()) && EmptyUtil.isNotEmpty(config.getMobile()) && ValidationUtil.isMobileNumberValid(config.getMobile())) {
            view.setMobile(config.getMobile());
        }
        view.setButtonText("Pay Rs " + StringUtil.formatNumber(NumberUtil.convertToRupees(config.getAmount())));
        view.showBranding();

        compositeSubscription.add(view.setImageClickListener().subscribe(aVoid -> {
            clicks += 1;
            if (clicks > 2) {
                clicks = 0;
                view.showSlogan();
            }
        }));
        HashMap<String, Observable<CharSequence>> map = view.setEditTextListener();
        compositeSubscription.add(map.get("mobile").subscribe(charSequence -> {
            view.setEditTextError("mobile", null);
            if (view.getPayButtonText().toLowerCase().contains("confirm")) {
                view.toggleConfirmationLayout(false);
            }
        }));
        compositeSubscription.add(map.get("code").subscribe(charSequence -> view.setConfirmationLayoutHeight("code")));
        compositeSubscription.add(map.get("pin").subscribe(charSequence -> view.setConfirmationLayoutHeight("pin")));
        compositeSubscription.add(view.setButtonClickListener().subscribe(aVoid -> {
            HashMap<String, String> dataMap = view.getFormData();
            if (view.getPayButtonText().toLowerCase().contains("confirm")) {
                if (isFinalFormValid(dataMap.get("pin"), dataMap.get("code"))) {
                    confirmPayment(view.hasNetwork(), dataMap.get("code"), dataMap.get("pin"));
                } else {
                    updateConfirmationHeight();
                }
            } else {
                if (isMobileValid(dataMap.get("mobile"))) {
                    initiatePayment(view.hasNetwork(), dataMap.get("mobile"));
                }
            }
        }));
    }

    @Override
    public void onDestroy() {
        if (EmptyUtil.isNotNull(compositeSubscription) && compositeSubscription.hasSubscriptions() && !compositeSubscription.isUnsubscribed()) {
            compositeSubscription.unsubscribe();
        }
        model.unSubscribe();
    }

    @Override
    public void openKhaltiSettings() {
        view.openKhaltiSettings();
    }

    @Override
    public void openLinkInBrowser() {
        view.openLinkInBrowser(Constant.url + pinWebLink.substring(1));
    }

    @Override
    public void showPINInBrowserDialog(String title, String message) {
        view.showPINInBrowserDialog(title, message);
    }

    @Override
    public boolean isMobileValid(String mobile) {
        if (EmptyUtil.isNotEmpty(mobile) && ValidationUtil.isMobileNumberValid(mobile)) {
            return true;
        } else {
            if (EmptyUtil.isEmpty(mobile)) {
                view.setEditTextError("mobile", "This field is required");
            } else {
                view.setEditTextError("mobile", "Enter a valid mobile number");
            }
        }
        return false;
    }

    @Override
    public boolean isFinalFormValid(String pin, String confirmationCode) {
        String status = "";
        if (EmptyUtil.isNotEmpty(pin)) {
            if (pin.length() == 4) {
                status += "pc";
            } else {
                status += "pl";
            }
        } else {
            status += "pe";
        }

        if (EmptyUtil.isNotEmpty(confirmationCode)) {
            if (confirmationCode.length() == 6) {
                status += "cc";
            } else {
                status += "cl";
            }
        } else {
            status += "ce";
        }

        switch (status) {
            case "pccc":
                return true;
            case "pccl":
                view.setEditTextError("code", "Enter a valid 6 digit confirmation code");
                return false;
            case "pcce":
                view.setEditTextError("code", "This field is required");
                return false;
            case "plcc":
                view.setEditTextError("pin", "Enter a valid 4 digit PIN");
                return false;
            case "plcl":
                view.setEditTextError("pin", "Enter a valid 4 digit PIN");
                view.setEditTextError("code", "Enter a valid 6 digit confirmation code");
                return false;
            case "plce":
                view.setEditTextError("pin", "Enter a valid 4 digit PIN");
                view.setEditTextError("code", "This field is required");
                return false;
            case "pecc":
                view.setEditTextError("pin", "This field is required");
                return false;
            case "pecl":
                view.setEditTextError("pin", "This field is required");
                view.setEditTextError("code", "Enter a valid 6 digit confirmation code");
                return false;
            case "pece":
                view.setEditTextError("pin", "This field is required");
                view.setEditTextError("code", "This field is required");
                return false;
        }
        return false;
    }

    @Override
    public void initiatePayment(boolean isNetwork, String mobile) {
        if (isNetwork) {
            view.toggleProgressDialog("init", true);
            compositeSubscription.add(model.initiatePayment(mobile, config)
                    .subscribe(new Subscriber<WalletInitPojo>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            view.toggleProgressDialog("init", false);
                            if (e.getMessage().contains("</a>")) {
                                pinWebLink = HtmlUtil.getHrefLink(e.getMessage());
                                view.showPINDialog("Error", view.getMessage("pin_not_set") + "\n\n" +
                                        view.getMessage("pin_not_set_continue"));
                                config.getOnCheckOutListener().onError(ErrorAction.WALLET_INITIATE.getAction(), view.getMessage("pin_not_set"));
                            } else {
                                view.showMessageDialog("Error", e.getMessage());
                                config.getOnCheckOutListener().onError(ErrorAction.WALLET_INITIATE.getAction(), e.getMessage());
                            }
                        }

                        @Override
                        public void onNext(WalletInitPojo walletInitPojo) {
                            view.toggleProgressDialog("init", false);
                            view.toggleConfirmationLayout(true);
                            view.togglePinMessage(walletInitPojo.isPinCreated());
                            view.setPinMessage(walletInitPojo.getPinCreatedMessage());
                        }
                    }));
        } else {
            view.showNetworkError();
        }
    }

    @Override
    public void confirmPayment(boolean isNetwork, String confirmationCode, String transactionPin) {
        if (isNetwork) {
            view.toggleProgressDialog("confirm", true);
            compositeSubscription.add(model.confirmPayment(confirmationCode, transactionPin)
                    .subscribe(new Subscriber<WalletConfirmPojo>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            view.toggleProgressDialog("confirm", false);
                            view.showMessageDialog("Error", e.getMessage());
                            config.getOnCheckOutListener().onError(ErrorAction.WALLET_CONFIRM.getAction(), e.getMessage());
                        }

                        @Override
                        public void onNext(WalletConfirmPojo walletConfirmPojo) {
                            view.toggleProgressDialog("confirm", false);
                            OnCheckOutListener onCheckOutListener = config.getOnCheckOutListener();
                            HashMap<String, Object> data = new HashMap<>();
                            data.putAll((EmptyUtil.isNotNull(config.getAdditionalData()) && EmptyUtil.isNotEmpty(config.getAdditionalData())) ? config.getAdditionalData() : new HashMap<>());
                            data.put("amount", walletConfirmPojo.getAmount());
                            data.put("product_url", walletConfirmPojo.getProductUrl());
                            data.put("token", walletConfirmPojo.getToken());
                            data.put("product_name", walletConfirmPojo.getProductName());
                            data.put("product_identity", walletConfirmPojo.getProductIdentity());

                            onCheckOutListener.onSuccess(data);
                            view.closeWidget();
                        }
                    }));
        } else {
            view.showNetworkError();
        }
    }

    @Override
    public void updateConfirmationHeight() {
        view.updateConfirmationHeight();
    }

    @Override
    public void unSubscribe() {
        if (compositeSubscription.hasSubscriptions() && !compositeSubscription.isUnsubscribed()) {
            compositeSubscription.unsubscribe();
        }
    }

    public void injectModel(WalletModel walletModel) {
        this.model = walletModel;
    }

    public void injectConfig(Config config) {
        this.config = config;
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/wallet/WalletPresenter.kt
package com.khalti.checkOut.wallet

import com.khalti.checkOut.api.ErrorAction
import com.khalti.checkOut.api.Result
import com.khalti.checkOut.helper.Config
import com.khalti.checkOut.wallet.helper.WalletInitPojo
import com.khalti.signal.CompositeSignal
import com.khalti.utils.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class WalletPresenter(view: WalletContract.View) : WalletContract.Presenter {

    /*Error messages*/
    private val EMPTY_ERROR = "This field is required"
    private val MOBILE_ERROR = "Enter a valid mobile number"
    private val PIN_ERROR = "Enter valid 4 digit PIN"
    private val CODE_ERROR = "Enter valid 6 digit confirmation code"

    /*Validation status*/
    private val MOBILE_VALID = "mv"
    private val PIN_VALID = "pv"
    private val CODE_VALID = "cv"

    private val MOBILE_EMPTY = "me"
    private val PIN_EMPTY = "pe"
    private val CODE_EMPTY = "ce"

    private val MOBILE_INVALID = "mi"
    private val PIN_INVALID = "pi"
    private val CODE_INVALID = "ci"

    private val view: WalletContract.View = GuavaUtil.checkNotNull<WalletContract.View>(view)
    private lateinit var config: Config
    private val compositeSignal = CompositeSignal()
    private val model = WalletModel()

    private val parentJob = Job()
    private val scope = CoroutineScope(Dispatchers.Main + parentJob)

    private var pinWebLink: String? = null
    private var clicks = 0

    private var mobile = ""

    private lateinit var walletInitPojo: WalletInitPojo

    init {
        view.setPresenter(this)
    }

    override fun onCreate() {
        config = Store.getConfig()
        val mobile = config.mobile
        if (EmptyUtil.isNotNull(mobile) && EmptyUtil.isNotEmpty(mobile) && ValidationUtil.isMobileNumberValid(mobile)) {
            view.setMobile(mobile!!)
        }

        view.setButtonText("Pay Rs " + StringUtil.formatNumber(NumberUtil.convertToRupees(config.amount)))
        view.showBranding()

        val clickMap = view.setClickListener()
        if (EmptyUtil.isNotNull(clickMap["pay"])) {
            compositeSignal.add(clickMap.getValue("pay").connect {

                val dataMap = view.formData
                if (view.payButtonText.toLowerCase().contains("confirm")) {
                    if (isFinalFormValid(dataMap.getValue("pin"), dataMap.getValue("code"))) {
                        onConfirmPayment(view.hasNetwork(), dataMap.getValue("code"), dataMap.getValue("pin"))
                    } else {
                        view.updateConfirmationHeight()
                    }
                } else {
                    if (isInitialFormValid(dataMap.getValue("mobile"), dataMap.getValue("pin"))) {
                        onInitiatePayment(view.hasNetwork(), dataMap.getValue("mobile"), dataMap.getValue("pin"))
                    }
                }
            })
        }

        if (EmptyUtil.isNotNull(clickMap["khalti"])) {
            compositeSignal.add(clickMap.getValue("khalti").connect {
                clicks += 1
                if (clicks > 2) {
                    clicks = 0
                    view.showSlogan()
                }
            })
        }

        val watcherMap = view.setEditTextListener()
        if (EmptyUtil.isNotNull(watcherMap["mobile"])) {
            compositeSignal.add(watcherMap.getValue("mobile")
                    .connect {
                        view.setEditTextError("mobile", null)
                        if (view.payButtonText.toLowerCase().contains("confirm")) {
                            view.toggleConfirmationLayout(false)
                        }
                    })
        }

        if (EmptyUtil.isNotNull(watcherMap["code"])) {
            compositeSignal.add(watcherMap.getValue("code")
                    .connect { view.setConfirmationLayoutHeight("code") })
        }

        if (EmptyUtil.isNotNull(watcherMap["pin"])) {
            compositeSignal.add(watcherMap.getValue("pin")
                    .connect { view.setConfirmationLayoutHeight("pin") })
        }
    }

    override fun onDestroy() {
        compositeSignal.clear()
        parentJob.cancel()
    }

    override fun isMobileValid(mobile: String): Boolean {
        if (EmptyUtil.isNotEmpty(mobile) && ValidationUtil.isMobileNumberValid(mobile)) {
            return true
        } else {
            if (EmptyUtil.isEmpty(mobile)) {
                view.setEditTextError("mobile", "This field is required")
            } else {
                view.setEditTextError("mobile", "Enter a valid mobile number")
            }
        }
        return false
    }

    override fun isInitialFormValid(mobile: String, pin: String): Boolean {
        var mobileError: String? = null
        var pinError: String? = null

        if (EmptyUtil.isNotEmpty(mobile)) {
            if (!ValidationUtil.isMobileNumberValid(mobile)) {
                mobileError = MOBILE_ERROR
            }
        } else {
            mobileError = EMPTY_ERROR
        }

        if (EmptyUtil.isNotEmpty(pin)) {
            if (pin.length < 4) {
                pinError = PIN_ERROR
            }
        } else {
            pinError = EMPTY_ERROR
        }

        view.setEditTextError("mobile", mobileError)
        view.setEditTextError("pin", pinError)

        return EmptyUtil.isNull(mobileError) && EmptyUtil.isNull(pinError)
    }

    override fun isFinalFormValid(pin: String, confirmationCode: String): Boolean {
        var status = ""
        status += if (EmptyUtil.isNotEmpty(pin)) {
            if (pin.length == 4) {
                "pc"
            } else {
                "pl"
            }
        } else {
            "pe"
        }

        status += if (EmptyUtil.isNotEmpty(confirmationCode)) {
            if (confirmationCode.length == 6) {
                "cc"
            } else {
                "cl"
            }
        } else {
            "ce"
        }

        when (status) {
            "pccc" -> return true
            "pccl" -> {
                view.setEditTextError("code", "Enter a valid 6 digit confirmation code")
                return false
            }
            "pcce" -> {
                view.setEditTextError("code", "This field is required")
                return false
            }
            "plcc" -> {
                view.setEditTextError("pin", "Enter a valid 4 digit PIN")
                return false
            }
            "plcl" -> {
                view.setEditTextError("pin", "Enter a valid 4 digit PIN")
                view.setEditTextError("code", "Enter a valid 6 digit confirmation code")
                return false
            }
            "plce" -> {
                view.setEditTextError("pin", "Enter a valid 4 digit PIN")
                view.setEditTextError("code", "This field is required")
                return false
            }
            "pecc" -> {
                view.setEditTextError("pin", "This field is required")
                return false
            }
            "pecl" -> {
                view.setEditTextError("pin", "This field is required")
                view.setEditTextError("code", "Enter a valid 6 digit confirmation code")
                return false
            }
            "pece" -> {
                view.setEditTextError("pin", "This field is required")
                view.setEditTextError("code", "This field is required")
                return false
            }
        }
        return false
    }

    override fun onInitiatePayment(isNetwork: Boolean, mobile: String, pin: String) {
        if (isNetwork) {
            this.mobile = mobile
            scope.launch {
                view.toggleProgressDialog("init", true)
                when (val result = model.initiatePayment(mobile, pin, config)) {
                    is Result.Success -> {
                        walletInitPojo = result.data
                        view.toggleProgressDialog("init", false)
                        view.toggleConfirmationLayout(true)
                        view.togglePinMessage(walletInitPojo.isPinCreated)
                        if (EmptyUtil.isNotNull(walletInitPojo.pinCreatedMessage)) {
                            view.setPinMessage(walletInitPojo.pinCreatedMessage!!)
                        }
                    }
                    is Result.Error -> {
                        view.toggleProgressDialog("init", false)
                        var message = result.throwable.message
                        if (EmptyUtil.isNotNull(message) && message!!.contains("</a>")) {
                            pinWebLink = HtmlUtil.getHrefLink(message)

                            compositeSignal.add(view.showPINDialog("Error", view.getMessage("pin_not_set") + "\n\n" + view.getMessage("pin_not_set_continue"))
                                    .connect { it ->
                                        if (it) {
                                            if (view.doesPackageExist()) {
                                                view.openKhaltiSettings()
                                            } else {
                                                compositeSignal.add(view.showPINDialog("Error", view.getMessage("khalti_not_found") + "\n\n" + view.getMessage("set_pin_in_browser"))
                                                        .connect {
                                                            if (it && EmptyUtil.isNotNull(pinWebLink)) {

                                                                view.openLinkInBrowser(Constant.url + pinWebLink!!.substring(1))
                                                            }
                                                        })
                                            }
                                        }
                                    })
                            config.onCheckOutListener.onError(ErrorAction.WALLET_INITIATE.action, message)
                        } else {
                            if (EmptyUtil.isNull(message)) {
                                message = ""
                            }
                            view.showMessageDialog("Error", message!!)
                        }
                    }
                }
            }
        } else {
            view.showNetworkError()
        }
    }

    override fun onConfirmPayment(isNetwork: Boolean, confirmationCode: String, transactionPin: String) {
        if (isNetwork) {
            scope.launch {
                if (EmptyUtil.isNotNull(walletInitPojo) && EmptyUtil.isNotNull(walletInitPojo.token)) {
                    view.toggleProgressDialog("confirm", true)
                    when (val result = model.confirmPayment(confirmationCode, transactionPin, walletInitPojo.token!!)) {
                        is Result.Success -> {
                            view.toggleProgressDialog("confirm", false)
                            val data = HashMap<String, Any>()
                            if (EmptyUtil.isNotNull(config.additionalData) && EmptyUtil.isNotEmpty(config.additionalData)) {
                                val walletConfirmPojo = result.data
                                data.putAll(config.additionalData!!)
                                data["mobile"] = mobile
                                if (EmptyUtil.isNotNull(walletConfirmPojo.amount)) {
                                    data["amount"] = walletConfirmPojo.amount!!
                                }
                                if (EmptyUtil.isNotNull(walletConfirmPojo.productUrl)) {
                                    data["product_url"] = walletConfirmPojo.productUrl!!
                                }
                                if (EmptyUtil.isNotNull(walletConfirmPojo.token)) {
                                    data["token"] = walletConfirmPojo.token!!
                                }
                                if (EmptyUtil.isNotNull(walletConfirmPojo.productName)) {
                                    data["product_name"] = walletConfirmPojo.productName!!
                                }
                                if (EmptyUtil.isNotNull(walletConfirmPojo.productIdentity)) {
                                    data["product_identity"] = walletConfirmPojo.productIdentity!!
                                }

                                config.onCheckOutListener.onSuccess(data)
                                view.closeWidget()
                            }

                        }

                        is Result.Error -> {
                            var message = result.throwable.message
                            if (EmptyUtil.isNull(message)) {
                                message = ""
                            }
                            view.toggleProgressDialog("confirm", false)
                            view.showMessageDialog("Error", message!!)
                            config.onCheckOutListener.onError(ErrorAction.WALLET_CONFIRM.action, message)
                        }

                    }
                }
            }
        } else {
            view.showNetworkError()
        }
    }
}